File: ExampleObjc/ExampleObjc/GetTreatmentViewController.m
//
//  GetTreatmentViewController.m
//  ExampleObjc
//
//  Created by Javier L. Avrudsky on 27/11/2018.
//  Copyright © 2018 Split Software. All rights reserved.
//

#import "GetTreatmentViewController.h"
@import Split;

@interface GetTreatmentViewController ()

@property (weak, nonatomic) IBOutlet UITextField *splitNameField;
@property (weak, nonatomic) IBOutlet UITextField *matchingKeyField;
@property (weak, nonatomic) IBOutlet UITextField *bucketingKeyField;
@property (weak, nonatomic) IBOutlet UITextField *attributesField;

@property (weak, nonatomic) IBOutlet UILabel *versionLabel;
@property (weak, nonatomic) IBOutlet UILabel *resultLabel;
@property (weak, nonatomic) IBOutlet UIButton *evaluateButton;

@end

@implementation GetTreatmentViewController

- (void)viewDidLoad {
    [super viewDidLoad];
}

- (IBAction)evaluateDidTouch:(UIButton *)sender {
    
    NSString *apiKey = @"YOUR_API_KEY";
    NSString *splitName = self.splitNameField.text;
    NSString *matchingKey = nil;
    
    if(self.matchingKeyField.text != nil &&
       ![[self.matchingKeyField.text stringByTrimmingCharactersInSet:NSCharacterSet.whitespaceAndNewlineCharacterSet] isEqualToString:@""]) {
        matchingKey = self.matchingKeyField.text;
    }
    
    // Split Config
    SplitClientConfig *config = [[SplitClientConfig alloc] init];
    config.featuresRefreshRate = 30;
    config.segmentsRefreshRate = 30;
    config.impressionRefreshRate = 30;
    config.sdkReadyTimeOut = 15000;
    config.connectionTimeout = 50;

    
    // Impression listener
    config.impressionListener = ^(Impression *impression){
        NSLog(@"Impression Key: %@", impression.keyName);
        NSLog(@"Impression Treatment: %@", impression.treatment);
        NSLog(@"Impression Time: %@", impression.timestamp);
        NSLog(@"Impression Change Number: %@", impression.changeNum);
    };
    
    //User Key
    Key *key = [[Key alloc] initWithMatchingKey:matchingKey bucketingKey:nil];
    
    //Split Factory
    SplitFactory *factory = [[SplitFactory alloc] initWithApiKey: apiKey key: key config: config];
    
    //Showing sdk version in UI
    self.versionLabel.text = [factory version];
    
    //Split Client
    id<SplitClientProtocol> client = [factory client];
    
    //Split Manager
    id<SplitManagerProtocol>manager = [factory manager];
    
    [client onEvent: SplitEventSdkReady execute: ^(){
        
        NSDictionary *attributes = [self convertToDictionary:self.attributesField.text];
        self.resultLabel.text = [client getTreatment:splitName attributes: attributes];
        
        // Get All Splits
        NSArray *splits = manager.splits;
        for(SplitView *split in splits) {
            NSLog(@"SplitView: %@, treatments: %@", split.name, [split.treatments componentsJoinedByString:@","]);
        }
        
        // Get Splits Names
        NSArray *splitNames = manager.splitNames;
        for(SplitView *splitName in splitNames) {
            NSLog(@"Split Name: %@", splitName);
        }
        
        // Find a Split
        if(splitNames.count > 0) {
            SplitView *split = [manager splitWithFeatureName:splitNames[0]];
            NSLog(@"Found: %@, treatments: %@", split.name, [split.treatments componentsJoinedByString:@","]);
        }
    }];
    
    [client onEvent: SplitEventSdkReadyTimedOut execute: ^(){
        self.resultLabel.text = @"SDK Time Out";
    }];
}

- ( NSDictionary* _Nullable ) convertToDictionary:(NSString*) text {
    NSData *data = [text dataUsingEncoding: kCFStringEncodingUTF8];
    if( data == nil) return nil;
    NSError *error = nil;
    NSDictionary *jsonObject = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&error];
    if(error != nil) {
        NSLog(@"Error parsing attributes: %@", error.localizedDescription);
        return nil;
    }
    return jsonObject;
}

@end

File: Example/Split/GetTreatmentViewController.swift
//
//  ViewController.swift
//  Split
//
//  Created by Brian Sztamfater on 09/18/2017.
//  Copyright (c) 2017 Split Software. All rights reserved.
//

import UIKit
import Split

class GetTreatmentViewController: UIViewController {
    
    @IBOutlet weak var bucketkey: UITextField?
    @IBOutlet weak var splitName: UITextField?
    @IBOutlet weak var matchingKey: UITextField?
    @IBOutlet weak var evaluteButton: UIButton?
    @IBOutlet weak var treatmentResult: UILabel?
    @IBOutlet weak var param1: UITextField?
    @IBOutlet weak var sdkVersion: UILabel?
    @IBOutlet weak var evaluateActivityIndicator: UIActivityIndicatorView!
    
    var factory: SplitFactory?
    var client: SplitClientProtocol?
    var manager: SplitManagerProtocol?

    @IBAction func evaluate(_ sender: Any) {
        evaluate()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        let tap = UITapGestureRecognizer(target: self, action: #selector(handleTap))
        self.view.addGestureRecognizer(tap)
    }
    
    @objc func handleTap(){
        self.view.endEditing(true)
    }
    
    func evaluate() {
        // Your Split API-KEY - Change in Config.swift file
        let authorizationKey: String = "YOUR_API_KEY"
        
        //Provided keys from UI
        let matchingKeyText: String = (matchingKey?.text)!
        let bucketing: String? = bucketkey?.text
        
        //Split Configuration
        let config = SplitClientConfig()
        
        config.featuresRefreshRate = 30
        config.segmentsRefreshRate = 30
        config.impressionRefreshRate = 30
        config.sdkReadyTimeOut = 15000
        config.connectionTimeout = 50



        config.impressionListener = { impression in
            print("\(impression.keyName ?? "") - \(impression.treatment ?? "") - \(impression.label ?? "")")
            DispatchQueue.global().async {
                // Do some async stuff
            }
        }
        
        //User Key
        let key: Key = Key(matchingKey: matchingKeyText, bucketingKey: bucketing)
        
        //Split Factory
        self.factory = SplitFactory(apiKey: authorizationKey, key: key, config: config)
        
        //Split Client
        self.client = self.factory?.client()
        
        //Split Manager
        self.manager = self.factory?.manager()
        
        //Showing sdk version in UI
        self.sdkVersion?.text = "SDK Version: \(self.factory?.version() ?? "unknown") "
        
        guard let client = self.client else {
            return
        }
        
        self.isEvaluating(active: true)
        client.on(event: SplitEvent.sdkReady) {
            [unowned self, unowned client = client] in
            
            // Main thread stuff
            var attributes: [String:Any]?
            if let json = self.param1?.text {
                attributes = self.convertToDictionary(text: json)
            }
            
            // Evaluate one split
            let treatment = client.getTreatment((self.splitName?.text)!, attributes: attributes)
            self.treatmentResult?.text = treatment
            
            // Get All Splits
            if let manager = self.manager {
                let splits = manager.splits
                splits.forEach {
                    print("SplitView: \($0)")
                }
                
                // Get Splits Names
                let splitNames = manager.splitNames
                splitNames.forEach {
                    print("Split Name: \($0)")
                }
                
                // Find a Split
                if splitNames.count > 0 {
                    let split = manager.split(featureName: splitNames[0])!
                    print("Found Split: \(split)")
                }
            }
            
            self.isEvaluating(active: false)
            DispatchQueue.global().async {
                // Do some async stuff
            }
        }
        
        client.on(event: SplitEvent.sdkReadyTimedOut) {
            [unowned self, unowned client = client] in
            // Main thread stuff
            var attributes: [String:Any]?
            if let json = self.param1?.text {
                attributes = self.convertToDictionary(text: json)
            }
            
            let treatment = client.getTreatment((self.splitName?.text)!, attributes: attributes)
            self.treatmentResult?.text = treatment
            
            self.isEvaluating(active: false)
            DispatchQueue.global().async {
                // Do some async stuff
            }
        }
    }
    
    func convertToDictionary(text: String) -> [String: Any]? {
        if let data = text.data(using: .utf8) {
            do {
                return try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
            } catch {
                print(error.localizedDescription)
            }
        }
        return nil
    }
    
    func isEvaluating(active: Bool){
        if active {
            self.evaluateActivityIndicator.startAnimating()
        } else {
            self.evaluateActivityIndicator.stopAnimating()
        }
    }
}
